pkg/reconciler/core/logicalcluster/logicalcluster_controller.go

```diff
@@ -0,0 +1,187 @@
+		logicalClusterIndexer: logicalClusterInformer.Informer().GetIndexer(),
+		logicalClusterLister:  logicalClusterInformer.Lister(),
+		commit:                committer.NewCommitter[*corev1alpha1.LogicalCluster, corev1alpha1client.LogicalClusterInterface, *corev1alpha1.LogicalClusterSpec, *corev1alpha1.LogicalClusterStatus](kcpClusterClient.CoreV1alpha1().LogicalClusters()),
+	}
+	logicalClusterInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueue(obj) },
+		UpdateFunc: func(obj, _ interface{}) { c.enqueue(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueue(obj) },
+	})
+
+	return c, nil
+}
+
+type logicalClusterResource = committer.Resource[*corev1alpha1.LogicalClusterSpec, *corev1alpha1.LogicalClusterStatus]
+
+// Controller watches Workspaces and WorkspaceShards in order to make sure every ClusterWorkspace
+// is scheduled to a valid Shard.
+type Controller struct {
+	queue workqueue.RateLimitingInterface
+
+	shardExternalURL func() string
+
+	kcpClusterClient kcpclientset.ClusterInterface
+
+	logicalClusterIndexer cache.Indexer
+	logicalClusterLister  corev1alpha1listers.LogicalClusterClusterLister
+
+	// commit creates a patch and submits it, if needed.
+	commit func(ctx context.Context, new, old *logicalClusterResource) error
+}
+
+func (c *Controller) enqueue(obj interface{}) {
+	key, err := kcpcache.MetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(2).Info("queueing LogicalCluster")
+	c.queue.Add(key)
+}
+
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+	}
+
+	<-ctx.Done()
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(1).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if requeue, err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", ControllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	} else if requeue {
+		// only requeue if we didn't error, but we still want to requeue
+		c.queue.Add(key)
```

Comment:
```suggestion
		c.queue.Add(key)
		return true
```

Reply:
done. Also fixed another instance in the workspace controller.